### PR TITLE
go/worker/keymanager/p2p: Remove retries and sticky peers

### PR DESCRIPTION
**Pull request by kostko**
Since retry and peer selection is now handled in the runtimes having it also done outside is detrimental to latency. The runtime knows better when to actually retry and which peers to select.

### CODE REVIEW

File: runtime/src/enclave_rpc/session.rs [hunk 1]
```diff
@@ -180,14 +185,15 @@ impl Session {
...
-        if let State::Transport(ref mut state) = self.state {
-            let msg = cbor::to_vec(msg);
-            let len = state.write_message(&msg, &mut self.buf)?;
-            writer.write_all(&self.buf[..len])?;
-
-            Ok(())
-        } else {
-            Err(SessionError::InvalidState.into())
+        match self.state {
+            State::Transport(ref mut state) | State::Abort(ref mut state) => {
```
Comment (peternose) [inline runtime/src/enclave_rpc/session.rs:189]: I would allow only `types::Message::Close` messages in abort state.

File: runtime/src/enclave_rpc/client.rs [hunk 2]
```diff
@@ -303,25 +309,35 @@ impl Controller {
         cbor::from_slice(&data).map_err(RpcClientError::DecodeError)
     }
 
-    async fn close(&mut self) -> Result<(), RpcClientError> {
-        if !self.session.inner.is_connected() {
-            return Ok(());
-        }
+    async fn reset(&mut self) {
+        // Notify the other end (if any) of session closure.
+        let _ = self.close_notify().await;
```
Comment (peternose) [inline runtime/src/enclave_rpc/client.rs:324]: This change added latency to the client whenever we do a reset.

File: runtime/src/enclave_rpc/session.rs [hunk 3]
```diff
@@ -132,32 +133,36 @@ impl Session {
...
-                if state.is_initiator() {
-                    // <- e, ee, s, es
-                    let len = state.read_message(&data, &mut self.buf)?;
-                    let remote_static = state
-                        .get_remote_static()
-                        .expect("dh exchange just happened");
-                    self.info = self
-                        .verify_rak_binding(&self.buf[..len], remote_static)
-                        .await?;
+                // <- e, ee, s, es (initiator) | <- s, se (responder)
```
Comment (peternose) [inline runtime/src/enclave_rpc/session.rs:136]:
```suggestion
```
Would remove as is distracting, or just mention that these were messages from handshake 1.

File: runtime/src/enclave_rpc/session.rs [hunk 4]
```diff
@@ -56,6 +56,7 @@ enum State {
     Handshake1(snow::HandshakeState),
     Handshake2(snow::HandshakeState),
     Transport(snow::TransportState),
+    Abort(snow::TransportState),
```
Comment (peternose) [inline runtime/src/enclave_rpc/session.rs:59]:
```suggestion
    UnauthenticatedTransport(snow::TransportState),
```

File: runtime/src/enclave_rpc/session.rs [hunk 5]
```diff
@@ -285,7 +291,7 @@ impl Session {
 
     /// Whether the session is in closed state.
     pub fn is_closed(&self) -> bool {
-        matches!(self.state, State::Closed)
+        matches!(self.state, State::Closed | State::Abort(_))
```
Comment (peternose) [inline runtime/src/enclave_rpc/session.rs:294]: Would split and add another method `is_unauthenticated`.

File: runtime/src/enclave_rpc/demux.rs [hunk 6]
```diff
@@ -15,13 +15,13 @@ use super::{
 use crate::{common::time::insecure_posix_system_time, identity::Identity};
 
 /// Maximum concurrent EnclaveRPC sessions.
-const MAX_CONCURRENT_SESSIONS: usize = 100;
+const MAX_CONCURRENT_SESSIONS: usize = 1024;
 /// Sessions without any processed frame for more than STALE_SESSION_TIMEOUT_SECS seconds
 /// can be purged.
-const STALE_SESSION_TIMEOUT_SECS: u64 = 60;
+const STALE_SESSION_TIMEOUT_SECS: u64 = 5;
```
Comment (peternose) [inline runtime/src/enclave_rpc/demux.rs:21]: Big changes. Should be tested thoroughly on testnet.

File: runtime/src/enclave_rpc/client.rs [hunk 7]
```diff
@@ -109,8 +109,16 @@ impl Controller {
                 Command::Call(request, kind, sender, retries) => {
                     self.call(request, kind, sender, retries).await
                 }
-                Command::PeerFeedback(pfid, peer_feedback) => {
+                Command::PeerFeedback(pfid, peer_feedback, kind) => {
                     self.transport.set_peer_feedback(pfid, Some(peer_feedback));
+
+                    // In case the peer feedback is bad, reset the session so a new peer can be
+                    // selected for a subsequent session.
+                    if !matches!(peer_feedback, types::PeerFeedback::Success)
```
Comment (peternose) [inline runtime/src/enclave_rpc/client.rs:117]: Feels like this doesn't belong here, if we even need this.

File: runtime/src/enclave_rpc/client.rs (hunk 2 above)
Comment (peternose) [reply runtime/src/enclave_rpc/client.rs:324]: RPC call timeout is 5 seconds, if the remove doesn't respond to close message.

File: runtime/src/enclave_rpc/client.rs (hunk 7 above)
Comment (kostko) [reply runtime/src/enclave_rpc/client.rs:117]: How will bad peer feedback from the application layer be taken into account then? As currently the node will be fixed unless a transport error happens.

File: runtime/src/enclave_rpc/demux.rs (hunk 6 above)
Comment (kostko) [reply runtime/src/enclave_rpc/demux.rs:21]: Will revert this part and add support for per peer limits instead.

File: runtime/src/enclave_rpc/client.rs (hunk 7 above)
Comment (peternose) [reply runtime/src/enclave_rpc/client.rs:117]: You are right, this is probably the way to go. However, I'm not sure if all errors on application layer should disconnect the underlying node. as connect calls are expensive (don't have an example for that, except when we replicate secrets, but there we switch to the next node if the current node returns an error). Should we let the caller decide that?

Comment (kostko) [reply runtime/src/enclave_rpc/client.rs:117]: Yeah the caller should decide, same as in the Go version.

Comment (kostko) [reply runtime/src/enclave_rpc/client.rs:117]: Added a way for the caller to provide feedback, please take a look.

File: runtime/src/enclave_rpc/demux.rs [hunk 8]
```diff
@@ -227,9 +356,9 @@ impl Demux {
             }
             Err(err) => {
                 // In case the session was closed, remove the session.
-                if session.session.is_closed() {
+                if session.inner.is_closed() {
                     let mut sessions = self.sessions.lock().unwrap();
-                    sessions.remove(&frame.session);
+                    sessions.remove(&mut session);
```
Comment (peternose) [inline runtime/src/enclave_rpc/demux.rs:361]:
```suggestion
                    sessions.remove(&session);
```

File: go/worker/keymanager/p2p/protocol.go [hunk 9]
```diff
@@ -16,10 +18,10 @@ const KeyManagerProtocolID = "keymanager"
 // KeyManagerProtocolVersion is the supported version of the keymanager protocol.
 var KeyManagerProtocolVersion = version.Version{Major: 2, Minor: 0, Patch: 0}
 
-// Constants related to the GetDiff method.
+// Constants related to the CallEnclave method.
 const (
-	MethodCallEnclave     = "CallEnclave"
-	MaxCallEnclaveRetries = 15
+	MethodCallEnclave          = "CallEnclave"
+	MaxCallEnclaveResponseTime = 3 * time.Second
```
Comment (peternose) [inline go/worker/keymanager/p2p/protocol.go:24]:
```suggestion
	MethodCallEnclaveTimeout = 3 * time.Second
```

File: .changelog/5546.internal.2.md [hunk 10]
```diff
@@ -0,0 +1 @@
+runtime/enclave_rpc: Support caller to provide peer feedback
```
Comment (peternose) [inline .changelog/5546.internal.2.md:1]: Why is this needed?

File: runtime/src/dispatcher.rs [hunk 11]
```diff
@@ -50,6 +51,15 @@ use crate::{
 /// Maximum amount of requests that can be in the dispatcher queue.
 const BACKLOG_SIZE: usize = 1000;
 
+/// Maximum number of concurrent peers. In case new peers are added, old sessions may be pruned.
+const RPC_MAX_PEERS: usize = 128;
+/// Maximum concurrent EnclaveRPC sessions per peer. In case more sessions are open, old
+/// sessions will be closed.
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/dispatcher.rs:57]: Of the peer that is trying to create a new session.

File: runtime/src/enclave_rpc/demux.rs [hunk 12]
```diff
@@ -57,164 +48,302 @@ impl From<Error> for crate::types::Error {
...
 
 /// A multiplexed session.
 pub struct MultiplexedSession {
-    id: SessionID,
-    session: Session,
-    last_process_frame_time: SystemTime,
+    peer_id: PeerID,
+    session_id: SessionID,
+    inner: Session,
+    last_process_frame_time: i64,
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/demux.rs:258]: Good idea, we can create an intermediate struct and store the access time there, avoiding the need to hold the async lock. Will change this.

File: runtime/src/dispatcher.rs [hunk 13]
```diff
@@ -50,6 +51,15 @@ use crate::{
...
+const RPC_MAX_SESSIONS_PER_PEER: usize = 8;
+/// Sessions without any processed frame for more than STALE_SESSION_TIMEOUT_SECS seconds
+/// can be purged.
+const RPC_STALE_SESSION_TIMEOUT_SECS: i64 = 60;
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/dispatcher.rs:61]: It was actually 60s before, but agree that it may be too much and will reduce to 10s.

File: runtime/src/enclave_rpc/demux.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/demux.rs]: Fixed.

File: runtime/src/enclave_rpc/demux.rs [hunk 14]
```diff
@@ -57,164 +48,302 @@ impl From<Error> for crate::types::Error {
...
+    by_peer: HashMap<PeerID, HashMap<SessionID, SharedSession>>,
+    /// A set of all sessions, ordered by idle time.
+    by_idle_time: BTreeSet<SessionByTimeKey>,
+}
+
+impl Sessions {
+    /// Create a new session management instance.
+    fn new(
+        builder: Builder,
+        max_peers: usize,
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/demux.rs]: Done.

File: runtime/src/enclave_rpc/session.rs [hunk 15]
```diff
@@ -491,7 +506,11 @@ impl Builder {
         let identity = self.identity.take();
         let remote_enclaves = self.remote_enclaves.take();
         let remote_runtime_id = self.remote_runtime_id.take();
-        let quote_policy = self.policy.take();
+        // If quote policy is set, use it, otherwise try to get it from identity if available.
+        let quote_policy = self
+            .policy
+            .take()
+            .or_else(|| identity.as_ref().and_then(|id| id.quote_policy()));
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/session.rs:513]: This does not change the existing behavior, it just moves the logic here instead of `create_session`. Let's do any changes in separate PRs.

File: runtime/src/enclave_rpc/demux.rs [hunk 16]
```diff
@@ -57,164 +48,293 @@ impl From<Error> for crate::types::Error {
...
+                    .iter()
+                    .min_by_key(|(_, s)| {
+                        if let Ok(_inner) = s.inner.try_lock() {
+                            s.last_access_time
+                        } else {
+                            i64::MAX // Session is currently in use.
+                        }
+                    })
+                    .map(|(sid, s)| (*sid, s.inner.clone(), s.by_time_key()))
+                    .unwrap();
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/demux.rs:170]: Actually I had this before, but I don't think it can panic as being here implies that at least one session exists for a peer and that one will be returned. But I guess it doesn't hurt to have this explicitly.

File: runtime/src/enclave_rpc/demux.rs [hunk 17]
```diff
@@ -57,164 +48,293 @@ impl From<Error> for crate::types::Error {
...
+                session.last_access_time = insecure_posix_time();
+                self.by_idle_time.insert(session.by_time_key());
+
+                return Ok((session.inner.clone(), false));
+            }
+
+            // Check if the peer has max sessions or if no more sessions are available globaly. If
+            // so, remove the oldest or return an error.
+            if sessions.len() >= self.max_sessions_per_peer
+                || self.by_idle_time.len() >= self.max_sessions
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/demux.rs:157]: Why? I think having this here makes more sense, e.g. imagine a peer has two out of four sessions open, but there are no more sessions globally. Without this a new connection would fail instead of evicting the other peer's session? Which I don't think is optimal?

File: runtime/src/enclave_rpc/demux.rs [hunk 18]
```diff
@@ -57,164 +48,293 @@ impl From<Error> for crate::types::Error {
...
+                        } else {
+                            i64::MAX // Session is currently in use.
+                        }
+                    })
+                    .map(|(sid, s)| (*sid, s.inner.clone(), s.by_time_key()))
+                    .unwrap();
+
+                if let Ok(_inner) = inner.try_lock() {
+                    sessions.remove(&session_id);
+                    self.by_idle_time.remove(&key);
```
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/demux.rs]: Changed to use `self.remove`.

File: runtime/src/enclave_rpc/session.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (peternose) [reply runtime/src/enclave_rpc/session.rs:513]: Ok, let's check in separate PR if this is true.
https://github.com/oasisprotocol/oasis-core/blob/master/keymanager/src/client/remote.rs#L156-L159

File: runtime/src/enclave_rpc/demux.rs [hunk 19]
```diff
@@ -57,164 +48,291 @@ impl From<Error> for crate::types::Error {
...
+                // Remove old idle time.
+                self.by_idle_time.remove(&session.by_time_key());
+                // Update idle time.
+                session.last_access_time = now;
+                self.by_idle_time.insert(session.by_time_key());
+
+                return Ok((session.inner.clone(), false));
+            }
+
+            // Check if the peer has max sessions or if no more sessions are available globaly. If
```
Comment (peternose) [inline runtime/src/enclave_rpc/demux.rs:154]:
```suggestion
            // Check if the peer has max sessions or if no more sessions are available globally. If
```

File: runtime/src/enclave_rpc/session.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/session.rs:513]: What you linked is for clients? This fallback to identity is used for demux/server as the quote policy is not set there.

[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/session.rs:513]: I guess we could also move this back to demux, but then you wouldn't be able to override for the server.

File: runtime/src/enclave_rpc/demux.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (peternose) [reply runtime/src/enclave_rpc/demux.rs:157]: No. If peer has 2/4, this `if` should be skipped, and the next `if` should take care to remove one global session.

File: runtime/src/enclave_rpc/session.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (peternose) [reply runtime/src/enclave_rpc/session.rs:513]: Fixed comment above (wrote quote instead of policy).

[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/session.rs:513]: Hm I guess we should only do this for responders but not initiators.

[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/session.rs:513]: Or actually maybe just handle this part in demux, that will be better.

File: runtime/src/enclave_rpc/demux.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/demux.rs:157]: But in that case there may be no sessions that are stale enough? While here we force remove the oldest one.

[the remark this answers is not in the document]
Comment (peternose) [reply runtime/src/enclave_rpc/demux.rs:157]: Not sure what is better. You can keep as it is.

File: runtime/src/enclave_rpc/session.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (kostko) [reply runtime/src/enclave_rpc/session.rs]: Fixed.

File: runtime/src/enclave_rpc/demux.rs [hunk 20]
```diff
@@ -57,164 +48,300 @@ impl From<Error> for crate::types::Error {
...
+
+    /// Create a new multiplexed session.
+    fn create_session(
+        mut builder: Builder,
+        peer_id: PeerID,
+        session_id: SessionID,
+        now: i64,
+    ) -> SessionMeta {
+        // If no quote policy is set, use the local one.
+        if builder.get_quote_policy().is_none() {
```
Comment (peternose) [inline runtime/src/enclave_rpc/demux.rs:122]: Previous version did not have this `if`. We always want to replace the policy, so that it is up to date. In this case, however, I think that the policy should be always `none`, so it's the same.

**Review by peternose — APPROVED** [review]
